### PR TITLE
KAFKA-16705 the flag "started" of RaftClusterInstance is false even though the cluster is started

### PR DESCRIPTION
**Pull request by gongxuanzhang**
fix KAFKA-16705

### CODE REVIEW

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 1]
```diff
@@ -284,24 +258,44 @@ public void startBroker(int brokerId) {
...
-            return clusterReference.get().controllers().values().stream();
+            return clusterTestKit.controllers().values().stream();
         }
 
+        private BrokerServer findBrokerOrThrow(int brokerId) {
+            return Optional.ofNullable(clusterTestKit.brokers().get(brokerId))
+                    .orElseThrow(() -> new IllegalArgumentException("Unknown brokerId " + brokerId));
+        }
+
+        private void buildAndFormatCluster() throws Exception {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java]: Could we add a phase `format` for kraft type? There are some test cases requiring `format`

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 2]
```diff
@@ -66,96 +65,66 @@ public class RaftClusterInvocationContext implements TestTemplateInvocationConte
...
-        this.zkReference = new AtomicReference<>();
         this.isCombined = isCombined;
     }
 
     @Override
     public String getDisplayName(int invocationIndex) {
         String clusterDesc = clusterConfig.nameTags().entrySet().stream()
-            .map(Object::toString)
-            .collect(Collectors.joining(", "));
+                .map(Object::toString)
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java]: please avoid those unrelated changes. smaller is better

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 3]
```diff
@@ -284,24 +258,51 @@ public void startBroker(int brokerId) {
...
-            return clusterReference.get().brokers().values().stream();
+            return clusterTestKit.brokers().values().stream();
         }
 
         public Stream<ControllerServer> controllers() {
-            return clusterReference.get().controllers().values().stream();
+            return clusterTestKit.controllers().values().stream();
         }
 
+        public void format() throws Exception {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:238]: `format` and `buildAndFormatCluster` can be merged. for example:
```java
        public void format() {
            if (this.clusterTestKit == null) {
                try {
                    KafkaClusterTestKit.Builder builder = new KafkaClusterTestKit.Builder(new TestKitNodes.Builder()
                        .setBootstrapMetadataVersion(clusterConfig.metadataVersion())
                        .setCombined(isCombined)
                        .setNumBrokerNodes(clusterConfig.numBrokers())
                        .setNumDisksPerBroker(clusterConfig.numDisksPerBroker())
                        .setPerServerProperties(clusterConfig.perServerOverrideProperties())
                        .setNumControllerNodes(clusterConfig.numControllers()).build());

                    if (Boolean.parseBoolean(clusterConfig.serverProperties()
                        .getOrDefault("zookeeper.metadata.migration.enable", "false"))) {
                        this.embeddedZookeeper = new EmbeddedZookeeper();
                        builder.setConfigProp("zookeeper.connect", String.format("localhost:%d", embeddedZookeeper.port()));
                    }

                    // Copy properties into the TestKit builder
                    clusterConfig.serverProperties().forEach(builder::setConfigProp);
                    // KAFKA-12512 need to pass security protocol and listener name here
                    this.clusterTestKit = builder.build();
                    this.clusterTestKit.format();
                } catch (Exception e) {
                    throw new RuntimeException("Failed to format Raft server", e);
                }
            }
        }
```

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 4]
```diff
@@ -252,7 +220,13 @@ public Admin createAdminClient(Properties configOverrides) {
         public void start() {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:178]: in this method we should always call `format` first. That is a big sugar to users

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 5]
```diff
@@ -171,39 +140,39 @@ public Optional<ListenerName> controllerListenerName() {
...
+                    .collect(Collectors.toList());
         }
 
         @Override
         public SocketServer anyBrokerSocketServer() {
             return brokers()
-                .map(BrokerServer::socketServer)
-                .findFirst()
-                .orElseThrow(() -> new RuntimeException("No broker SocketServers found"));
+                    .map(BrokerServer::socketServer)
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java]: ditto. please revert those changes.

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for this patch

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 6]
```diff
@@ -284,24 +259,59 @@ public void startBroker(int brokerId) {
...
+        private BrokerServer findBrokerOrThrow(int brokerId) {
+            return Optional.ofNullable(clusterTestKit.brokers().get(brokerId))
+                    .orElseThrow(() -> new IllegalArgumentException("Unknown brokerId " + brokerId));
+        }
+
+        private void safeBuildCluster() throws Exception {
+            if (this.clusterTestKit != null) {
+                return;
+            }
+            synchronized (this) {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java]: I don't think we need to sync, as it won't be called concurrent.

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 7]
```diff
@@ -284,24 +259,59 @@ public void startBroker(int brokerId) {
...
-            return clusterReference.get().brokers().values().stream();
+            return clusterTestKit.brokers().values().stream();
         }
 
         public Stream<ControllerServer> controllers() {
-            return clusterReference.get().controllers().values().stream();
+            return clusterTestKit.controllers().values().stream();
+        }
+
+        public void format() throws Exception {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:238]: This method is not idempotent, so user can re-format it if they call `format` and then `start`

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for updated PR

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 8]
```diff
@@ -240,28 +215,61 @@ public void startBroker(int brokerId) {
...
                     .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
         }
 
         @Override
         public Map<Integer, ControllerServer> controllers() {
-            return Collections.unmodifiableMap(clusterReference.get().controllers());
+            return Collections.unmodifiableMap(clusterTestKit.controllers());
+        }
+
+        public void format() throws Exception {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:238]: We can put `safeBuildCluster` and `doBuild` into `format`, right?
```java
        public void format() throws Exception {
            if (formated.compareAndSet(false,true)) {
                TestKitNodes nodes = new TestKitNodes.Builder()
                    .setBootstrapMetadataVersion(clusterConfig.metadataVersion())
                    .setCombined(isCombined)
                    .setNumBrokerNodes(clusterConfig.numBrokers())
                    .setNumDisksPerBroker(clusterConfig.numDisksPerBroker())
                    .setPerServerProperties(clusterConfig.perServerOverrideProperties())
                    .setNumControllerNodes(clusterConfig.numControllers()).build();
                KafkaClusterTestKit.Builder builder = new KafkaClusterTestKit.Builder(nodes);
                if (Boolean.parseBoolean(clusterConfig.serverProperties()
                    .getOrDefault("zookeeper.metadata.migration.enable", "false"))) {
                    this.embeddedZookeeper = new EmbeddedZookeeper();
                    builder.setConfigProp("zookeeper.connect", String.format("localhost:%d", embeddedZookeeper.port()));
                }
                // Copy properties into the TestKit builder
                clusterConfig.serverProperties().forEach(builder::setConfigProp);
                // KAFKA-12512 need to pass security protocol and listener name here
                this.clusterTestKit = builder.build();
                this.clusterTestKit.format();
            }
        }
```

**Comment by chia7712** [comment]
```
[2024-05-17T02:48:07.827Z] [ant:checkstyle] [ERROR] /home/jenkins/workspace/Kafka_kafka-pr_PR-15946/core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:238:45: ',' is not followed by whitespace. [WhitespaceAfter]

```

Could you please fix the build error?

**Comment by gongxuanzhang** [comment]
> Could you please fix the build error?

I can,I will submit later

**Comment by gongxuanzhang** [comment]
> Could you please fix the build error?

fixed it, i'm so sorry.

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 9]
```diff
@@ -86,67 +81,42 @@ public String getDisplayName(int invocationIndex) {
...
-                        () -> cluster.brokers().get(0).brokerState() == BrokerState.RUNNING,
-                        () -> "Broker never made it to RUNNING state.",
-                        org.apache.kafka.test.TestUtils.DEFAULT_MAX_WAIT_MS,
-                        100L);
-                }
-            },
-            (AfterTestExecutionCallback) context -> clusterInstance.stop(),
-            new ClusterInstanceParameterResolver(clusterInstance)
+                (BeforeTestExecutionCallback) context -> {
+                    if (clusterConfig.isAutoStart()) {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:88]: In order to keep compatibility (less changes to tests), we should add `clusterInstance.format();`

**Comment by gongxuanzhang** [comment]
| test | jira |
|-------|-------|
| testReplicateSourceDefault| https://issues.apache.org/jira/browse/KAFKA-15926 |
| testThreeCompressedRecordsInSeparateBatch| https://issues.apache.org/jira/browse/KAFKA-15731 |
| testTaskRequestWithOldStartMsGetsUpdated| https://issues.apache.org/jira/browse/KAFKA-15760 |
| testSyncTopicConfigs | https://issues.apache.org/jira/browse/KAFKA-15523 |
| testAlterSinkConnectorOffsetsOverriddenConsumerGroupId| https://issues.apache.org/jira/browse/KAFKA-15914 |
| testConsumptionWithBrokerFailures | https://issues.apache.org/jira/browse/KAFKA-15146 |
| testNoConsumeWithDescribeAclViaAssign| https://issues.apache.org/jira/browse/KAFKA-15411 |

**Review by chia7712 — APPROVED** [review]
